{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import scrublet as scr\n",
    "import scipy.io\n",
    "import matplotlib\n",
    "matplotlib.use('TkAgg')\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import os\n",
    "import pandas as pd\n",
    "import csv\n",
    "import scanpy as sc\n",
    "\n",
    "#gzip -dk out_*/filtered_feature_bc_matrix/*.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'4918STDY8615259'"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samples = pd.read_csv (\"../../../data/raw/scRNAseq/4918stdy_organoids.csv\")\n",
    "samples[\"SANGER SAMPLE ID\"][0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Counts matrix shape: 5737 rows, 33538 columns\n",
      "Number of feature in feature list: 33538\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.40\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.1%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 6.0%\n",
      "\tEstimated  = 0.0%\n",
      "Elapsed time: 5.5 seconds\n",
      "Counts matrix shape: 3864 rows, 33538 columns\n",
      "Number of feature in feature list: 33538\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.47\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 6.0%\n",
      "\tEstimated  = 0.0%\n",
      "Elapsed time: 3.2 seconds\n",
      "Counts matrix shape: 3697 rows, 33538 columns\n",
      "Number of feature in feature list: 33538\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.55\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.1%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 6.0%\n",
      "\tEstimated  = 20.0%\n",
      "Elapsed time: 3.0 seconds\n",
      "Counts matrix shape: 4151 rows, 33538 columns\n",
      "Number of feature in feature list: 33538\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.47\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.1%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 6.0%\n",
      "\tEstimated  = 0.0%\n",
      "Elapsed time: 4.0 seconds\n",
      "Counts matrix shape: 3078 rows, 33538 columns\n",
      "Number of feature in feature list: 33538\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.53\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.2%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 6.0%\n",
      "\tEstimated  = 0.0%\n",
      "Elapsed time: 2.6 seconds\n",
      "Counts matrix shape: 2977 rows, 33538 columns\n",
      "Number of feature in feature list: 33538\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.52\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.2%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 6.0%\n",
      "\tEstimated  = 0.0%\n",
      "Elapsed time: 2.5 seconds\n",
      "Counts matrix shape: 5047 rows, 33538 columns\n",
      "Number of feature in feature list: 33538\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.44\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.1%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 6.0%\n",
      "\tEstimated  = 0.0%\n",
      "Elapsed time: 4.4 seconds\n",
      "Counts matrix shape: 5664 rows, 33538 columns\n",
      "Number of feature in feature list: 33538\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.45\n",
      "Detected doublet rate = 0.1%\n",
      "Estimated detectable doublet fraction = 0.3%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 6.0%\n",
      "\tEstimated  = 25.8%\n",
      "Elapsed time: 5.1 seconds\n",
      "Counts matrix shape: 4823 rows, 33538 columns\n",
      "Number of feature in feature list: 33538\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.58\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.1%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 6.0%\n",
      "\tEstimated  = 0.0%\n",
      "Elapsed time: 4.4 seconds\n",
      "Counts matrix shape: 4074 rows, 33538 columns\n",
      "Number of feature in feature list: 33538\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.57\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.1%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 6.0%\n",
      "\tEstimated  = 0.0%\n",
      "Elapsed time: 3.9 seconds\n",
      "Counts matrix shape: 2681 rows, 33538 columns\n",
      "Number of feature in feature list: 33538\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.42\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.6%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 6.0%\n",
      "\tEstimated  = 6.2%\n",
      "Elapsed time: 2.1 seconds\n",
      "Counts matrix shape: 4171 rows, 33538 columns\n",
      "Number of feature in feature list: 33538\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.57\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 6.0%\n",
      "\tEstimated  = 0.0%\n",
      "Elapsed time: 3.7 seconds\n"
     ]
    }
   ],
   "source": [
    "doublet_threshold = []\n",
    "\n",
    "for samp in range(0, len(samples[\"SANGER SAMPLE ID\"])):\n",
    "    input_dir = '../../../data/raw/scRNAseq/scRNASEQ_DATA/Pediatric_organoids/out_'+samples[\"SANGER SAMPLE ID\"][samp]+'/filtered_feature_bc_matrix/'\n",
    "    counts_matrix = scipy.io.mmread(input_dir + 'matrix.mtx').T.tocsc()\n",
    "    genes = np.array(scr.load_genes(input_dir + 'features.tsv', delimiter='\\t', column=1))\n",
    "    \n",
    "    print('Counts matrix shape: {} rows, {} columns'.format(counts_matrix.shape[0], counts_matrix.shape[1]))\n",
    "    print('Number of feature in feature list: {}'.format(len(genes)))\n",
    "    \n",
    "    scrub = scr.Scrublet(counts_matrix, expected_doublet_rate=0.06)\n",
    "    \n",
    "    doublet_scores, predicted_doublets = scrub.scrub_doublets(min_counts=2, \n",
    "                                                          min_cells=3, \n",
    "                                                          min_gene_variability_pctl=85, \n",
    "                                                          n_prin_comps=30)\n",
    "    #save plot\n",
    "    scrub.plot_histogram()\n",
    "    plt.savefig('../../../figs/scRNAseq/doublets_organoid/doublet_histogram_' + samples[\"SANGER SAMPLE ID\"][samp] + '.pdf')\n",
    "    \n",
    "    #save thesholds to report\n",
    "    doublet_threshold.append(scrub.threshold_)\n",
    "    \n",
    "    #save output for R/seurat\n",
    "    barcodes = pd.read_csv (input_dir + 'barcodes.tsv', sep = '\\t', header=None)\n",
    "    \n",
    "    df = pd.DataFrame({\n",
    "    'index': barcodes[0],\n",
    "    'doublet_score': scrub.doublet_scores_obs_,\n",
    "    'predicted_doublet': scrub.predicted_doublets_\n",
    "    })\n",
    "    df.to_csv('../../../output/scrublet_output_table_' + samples[\"SANGER SAMPLE ID\"][samp] + '.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "doublet_threshold\n",
    "doublet_threshold_df = pd.DataFrame(doublet_threshold)\n",
    "doublet_threshold_df.to_csv(\"../../../output/doublet_threshold_organoid.csv\") "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
