{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "import scrublet as scr\n",
    "import scipy.io\n",
    "import matplotlib\n",
    "matplotlib.use('TkAgg')\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import os\n",
    "import pandas as pd\n",
    "import csv\n",
    "import scanpy as sc\n",
    "\n",
    "\n",
    "#gzip -dk outs_*/filtered_feature_bc_matrix/*.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'4918STDY7273964'"
      ]
     },
     "execution_count": 90,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samples = pd.read_csv (\"../../../data/raw/scRNAseq/scRNASEQ_DATA/Pediatric_samples/Pediatric_sample_tracker.csv\")\n",
    "samples[\"Sanger Sample ID\"][0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Counts matrix shape: 6459 rows, 33538 columns\n",
      "Number of feature in feature list: 33538\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.38\n",
      "Detected doublet rate = 0.9%\n",
      "Estimated detectable doublet fraction = 17.9%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 6.0%\n",
      "\tEstimated  = 4.9%\n",
      "Elapsed time: 5.3 seconds\n",
      "Counts matrix shape: 4993 rows, 33538 columns\n",
      "Number of feature in feature list: 33538\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.56\n",
      "Detected doublet rate = 0.2%\n",
      "Estimated detectable doublet fraction = 7.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 6.0%\n",
      "\tEstimated  = 2.6%\n",
      "Elapsed time: 3.9 seconds\n",
      "Counts matrix shape: 3582 rows, 33538 columns\n",
      "Number of feature in feature list: 33538\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.42\n",
      "Detected doublet rate = 0.5%\n",
      "Estimated detectable doublet fraction = 17.1%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 6.0%\n",
      "\tEstimated  = 2.8%\n",
      "Elapsed time: 2.8 seconds\n",
      "Counts matrix shape: 8895 rows, 33538 columns\n",
      "Number of feature in feature list: 33538\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.66\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 1.1%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 6.0%\n",
      "\tEstimated  = 4.0%\n",
      "Elapsed time: 7.7 seconds\n",
      "Counts matrix shape: 7173 rows, 33538 columns\n",
      "Number of feature in feature list: 33538\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.63\n",
      "Detected doublet rate = 0.1%\n",
      "Estimated detectable doublet fraction = 2.1%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 6.0%\n",
      "\tEstimated  = 4.7%\n",
      "Elapsed time: 5.8 seconds\n",
      "Counts matrix shape: 2791 rows, 33538 columns\n",
      "Number of feature in feature list: 33538\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.34\n",
      "Detected doublet rate = 0.8%\n",
      "Estimated detectable doublet fraction = 33.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 6.0%\n",
      "\tEstimated  = 2.3%\n",
      "Elapsed time: 1.8 seconds\n",
      "Counts matrix shape: 7026 rows, 33538 columns\n",
      "Number of feature in feature list: 33538\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.63\n",
      "Detected doublet rate = 0.1%\n",
      "Estimated detectable doublet fraction = 3.1%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 6.0%\n",
      "\tEstimated  = 2.3%\n",
      "Elapsed time: 5.4 seconds\n",
      "Counts matrix shape: 3350 rows, 33538 columns\n",
      "Number of feature in feature list: 33538\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.52\n",
      "Detected doublet rate = 0.1%\n",
      "Estimated detectable doublet fraction = 9.7%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 6.0%\n",
      "\tEstimated  = 0.9%\n",
      "Elapsed time: 1.9 seconds\n",
      "Counts matrix shape: 3662 rows, 33538 columns\n",
      "Number of feature in feature list: 33538\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.53\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 1.7%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 6.0%\n",
      "\tEstimated  = 1.7%\n",
      "Elapsed time: 2.4 seconds\n",
      "Counts matrix shape: 8537 rows, 33538 columns\n",
      "Number of feature in feature list: 33538\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.44\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.2%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 6.0%\n",
      "\tEstimated  = 15.4%\n",
      "Elapsed time: 6.9 seconds\n",
      "Counts matrix shape: 5427 rows, 33538 columns\n",
      "Number of feature in feature list: 33538\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.60\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.4%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 6.0%\n",
      "\tEstimated  = 4.3%\n",
      "Elapsed time: 3.8 seconds\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/redgar/anaconda3/envs/IBD_scRNAseq/lib/python3.8/site-packages/scrublet/scrublet.py:487: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).\n",
      "  fig, axs = plt.subplots(1, 2, figsize = fig_size)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Counts matrix shape: 2466 rows, 33538 columns\n",
      "Number of feature in feature list: 33538\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.49\n",
      "Detected doublet rate = 0.1%\n",
      "Estimated detectable doublet fraction = 11.5%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 6.0%\n",
      "\tEstimated  = 1.1%\n",
      "Elapsed time: 1.5 seconds\n",
      "Counts matrix shape: 1705 rows, 33538 columns\n",
      "Number of feature in feature list: 33538\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.45\n",
      "Detected doublet rate = 0.1%\n",
      "Estimated detectable doublet fraction = 10.9%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 6.0%\n",
      "\tEstimated  = 1.1%\n",
      "Elapsed time: 0.9 seconds\n",
      "Counts matrix shape: 4114 rows, 33538 columns\n",
      "Number of feature in feature list: 33538\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.37\n",
      "Detected doublet rate = 0.4%\n",
      "Estimated detectable doublet fraction = 24.8%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 6.0%\n",
      "\tEstimated  = 1.6%\n",
      "Elapsed time: 3.1 seconds\n",
      "Counts matrix shape: 13003 rows, 33538 columns\n",
      "Number of feature in feature list: 33538\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.56\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.1%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 6.0%\n",
      "\tEstimated  = 0.0%\n",
      "Elapsed time: 12.1 seconds\n",
      "Counts matrix shape: 7329 rows, 33538 columns\n",
      "Number of feature in feature list: 33538\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.31\n",
      "Detected doublet rate = 1.4%\n",
      "Estimated detectable doublet fraction = 19.1%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 6.0%\n",
      "\tEstimated  = 7.4%\n",
      "Elapsed time: 5.8 seconds\n",
      "Counts matrix shape: 3486 rows, 33538 columns\n",
      "Number of feature in feature list: 33538\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.34\n",
      "Detected doublet rate = 0.9%\n",
      "Estimated detectable doublet fraction = 24.5%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 6.0%\n",
      "\tEstimated  = 3.6%\n",
      "Elapsed time: 2.2 seconds\n",
      "Counts matrix shape: 3667 rows, 33538 columns\n",
      "Number of feature in feature list: 33538\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.37\n",
      "Detected doublet rate = 1.0%\n",
      "Estimated detectable doublet fraction = 22.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 6.0%\n",
      "\tEstimated  = 4.3%\n",
      "Elapsed time: 2.6 seconds\n",
      "Counts matrix shape: 1406 rows, 33538 columns\n",
      "Number of feature in feature list: 33538\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.42\n",
      "Detected doublet rate = 0.4%\n",
      "Estimated detectable doublet fraction = 17.2%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 6.0%\n",
      "\tEstimated  = 2.1%\n",
      "Elapsed time: 0.9 seconds\n",
      "Counts matrix shape: 6312 rows, 33538 columns\n",
      "Number of feature in feature list: 33538\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.29\n",
      "Detected doublet rate = 1.6%\n",
      "Estimated detectable doublet fraction = 34.2%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 6.0%\n",
      "\tEstimated  = 4.8%\n",
      "Elapsed time: 4.8 seconds\n",
      "Counts matrix shape: 5271 rows, 33538 columns\n",
      "Number of feature in feature list: 33538\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.30\n",
      "Detected doublet rate = 1.7%\n",
      "Estimated detectable doublet fraction = 31.4%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 6.0%\n",
      "\tEstimated  = 5.3%\n",
      "Elapsed time: 3.6 seconds\n",
      "Counts matrix shape: 2995 rows, 33538 columns\n",
      "Number of feature in feature list: 33538\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.49\n",
      "Detected doublet rate = 0.4%\n",
      "Estimated detectable doublet fraction = 10.6%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 6.0%\n",
      "\tEstimated  = 3.5%\n",
      "Elapsed time: 2.4 seconds\n"
     ]
    }
   ],
   "source": [
    "doublet_threshold = []\n",
    "\n",
    "for samp in range(0, len(samples[\"Sanger Sample ID\"])):\n",
    "    input_dir = '../../../data/raw/scRNAseq/scRNASEQ_DATA/Pediatric_samples/pediatric_cellranger302_GRCh38-3.0.0/outs_'+samples[\"Sanger Sample ID\"][samp]+'/filtered_feature_bc_matrix/'\n",
    "    counts_matrix = scipy.io.mmread(input_dir + 'matrix.mtx').T.tocsc()\n",
    "    genes = np.array(scr.load_genes(input_dir + 'features.tsv', delimiter='\\t', column=1))\n",
    "    \n",
    "    print('Counts matrix shape: {} rows, {} columns'.format(counts_matrix.shape[0], counts_matrix.shape[1]))\n",
    "    print('Number of feature in feature list: {}'.format(len(genes)))\n",
    "    \n",
    "    scrub = scr.Scrublet(counts_matrix, expected_doublet_rate=0.06)\n",
    "    \n",
    "    doublet_scores, predicted_doublets = scrub.scrub_doublets(min_counts=2, \n",
    "                                                          min_cells=3, \n",
    "                                                          min_gene_variability_pctl=85, \n",
    "                                                          n_prin_comps=30)\n",
    "    #save plot\n",
    "    scrub.plot_histogram()\n",
    "    plt.savefig('../../../figs/scRNAseq/primary/doublets/doublet_histogram_' + samples[\"Sanger Sample ID\"][samp] + '.pdf')\n",
    "    \n",
    "    #save thesholds to report\n",
    "    doublet_threshold.append(scrub.threshold_)\n",
    "    \n",
    "    #save output for R/seurat\n",
    "    barcodes = pd.read_csv (input_dir + 'barcodes.tsv', sep = '\\t', header=None)\n",
    "    \n",
    "    df = pd.DataFrame({\n",
    "    'index': barcodes[0],\n",
    "    'doublet_score': scrub.doublet_scores_obs_,\n",
    "    'predicted_doublet': scrub.predicted_doublets_\n",
    "    })\n",
    "    df.to_csv('../../../output/scrublet_output_table_' + samples[\"Sanger Sample ID\"][samp] + '.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {},
   "outputs": [],
   "source": [
    "doublet_threshold\n",
    "doublet_threshold_df = pd.DataFrame(doublet_threshold)\n",
    "doublet_threshold_df.to_csv(\"../../../output/doublet_threshold_primary.csv\") "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
